### ISSUE TITLE
Support for HTTP proxy

### ISSUE THREAD

**Issue by mrevutskyi** [Azure/azure-kusto-python#350]
**Is your feature request related to a problem? Please describe.**
My previous request #317 was closed without much explanation or examples, except that I _may be_ can use it.
I was not able to find any relevant changes or examples how to correctly pass proxy settings.

**Describe the solution you'd like**
An example or a link to a relevant code/parameter that would allow me to use HTTP proxy with KustoClient

**Comment by ArielYehezkely** [comment]
Hi @mrevutskyi 

Please see the attached PR - will it satisfy your need?

**Comment by mrevutskyi** [comment]
Hi @ArielYehezkely 
thank you for your response.

Proxy setting also need to be passed to the [auth provider](https://github.com/Azure/azure-kusto-python/blob/8a6d654ef2862190d037dff8b2a08151fb040997/azure-kusto-data/azure/kusto/data/client.py#L772) 
Otherwise the client will not be able to authenticate

**Comment by mrevutskyi** [comment]
if it helps, here is a snippet of how we work around this now
```python

import adal
from azure.kusto.data import KustoClient, KustoConnectionStringBuilder
from azure.kusto.data.security import _AadHelper as AadHelper

class ProxiedAadHelper(AadHelper):
    def __init__(self, kcsb: KustoConnectionStringBuilder, proxies: Dict[str, str]):
        super().__init__(kcsb)
        self.auth_context = adal.AuthenticationContext(f'https://login.microsoftonline.com/{kcsb.authority_id}', proxies=proxies)


class ProxiedKustoClient(KustoClient):
    def __init__(self, kcsb: KustoConnectionStringBuilder, proxy: str):
        super().__init__(kcsb)
        proxy_configs = {'http': proxy, 'https': proxy}
        self._auth_provider = ProxiedAadHelper(kcsb, proxy_configs)
        self._session.proxies.update(proxy_configs)
```   

The problem is that `_AadHelper`,  `self._auth_provider`, `self._session` are all protected/private, so we can not rely on them, to be present/consistent even in minor version updates

**Comment by yogilad** [comment]
@mrevutskyi , what the scenario here.
I'm not very eager to allow proxing the call to the authentication calls as they include credentials.
You can always provide you own authentication implementation using with_add_token_provider which gets a callback to a token generating function.

@AsafMah , can you issue a fix for the SDK which disables the call to cloud settings where not needed (Specifically for token provider)?

**Comment by mrevutskyi** [comment]
The scenario is very common: service is running inside a DMZ network and Internet is only accessible via secure proxy. It still uses SSL, so credentials are safe.

I get an impression that there is a misunderstanding about what kind of `proxy` is requested. I'm talking about HTTP/HTTPS proxy, not a proxy class or a callback. 

Currently there is no suitable interface to implement our own authentication. For example `_auth_provider` is a private property in `KustoClient`. We could use our existing solution posted above if `_AadHelper` become public and there was a public interface for using custom auth provider.

**Comment by yogilad** [comment]
Have a look at `KustoConnectionStringBuilder.with_token_provider(cls, connection_string: str, token_provider: Callable[[], str]) -> "KustoConnectionStringBuilder":`

TokenProvider is a function returning the token string to use.

**Comment by AsafMah** [comment]
It will probably fail at the cloud info until we merge https://github.com/Azure/azure-kusto-python/pull/357

**Comment by yogilad** [comment]
Additionally, the _AadHelper overriding approach is not the one I'd use.
The proxy needs to be set in all MSAL based TokenProviders, and in CloudSettings.

@ArielYehezkely , this could be done by the client passing it's proxy to the exiting auth code.

**Comment by mrevutskyi** [comment]
That's confusing, 
we use `KustoConnectionStringBuilder.with_aad_application_key_authentication`, are you suggesting that we need to reimplement it as a token provider function?

I agree that `_AadHelper` is not the best approach, hoping to find a better solution with this request

**Comment by yogilad** [comment]
@mrevutskyi, Yes - you can technically do that if you don't want to wait for a fix.

**Comment by AsafMah** [comment]
Finally closed with #360
